# Find Meijer store locations.

Use the Meijer API to return a list of all stores.

## Lessons Learned.

- Using [Fake GPS location](https://play.google.com/store/apps/details?id=com.lexa.fakegps&hl=en_US) Meijer's app searches for location (0.0, 0.0).

- API URL: https://mservices.meijer.com/storeinfo/api/mobile/near?latitude=0.0&longitude=0.0&miles=2000&numToReturn=250

Fred Meijer Offices:

2929 Walker Ave NW, Grand Rapids, MI 49544

43.016570, -85.726290

In [1]:
latitude, longitude=43.016570, -85.726290

In [3]:
import os
import requests
import base64
request=dict()

account_services_secret="drAqas76Re7RekeBanaMaNEMah7paDE5"
basic_token=base64.encodebytes("mma:{}".format(account_services_secret).encode("UTF-8")).decode("UTF-8").strip()

request["url"] = "https://mservices.meijer.com/storeinfo/api/mobile/near"
request["headers"] = {
    'Authorization': 'Basic {}'.format(basic_token),
    'Platform': 'Android',
    'Version': '5.10.0',
    'Build': '51000000',
    'Connection': "Keep-Alive",
    'Accept-Encoding': 'gzip',
    'User-Agent': 'okhttp/3.8.0'
}
request["params"] =  {
    'latitude': latitude,
    'longitude': longitude,
    'miles': '10000',
    'numToReturn': '10000',
}
r = requests.get(**request)
r

<Response [401]>

In [7]:
headers = {
    "Authorization": os.environ["MEIJER_AUTH"],
    "Version": "6",
    "User-Agent": "okhttp/3.8.0",
}

KeyError: 'MEIJER_AUTH'

In [86]:
url = 'https://mservices.meijer.com/storeinfo/api/mobile/near'
payload = {
    'latitude': Latitude,
    'longitude': Longitude,
    'miles': '10000',
    'numToReturn': '10000',
}

In [87]:
r = requests.get(url, params=payload, headers=headers)

In [88]:
r.status_code

200

In [89]:
len(r.json()["store"])

242

From Wiki: https://en.wikipedia.org/wiki/Meijer.

> About half of the company's **242 stores** are located in Michigan, with additional locations in Illinois, Indiana, Kentucky, Ohio, and Wisconsin.

In [97]:
(r.json()).keys()

dict_keys(['store'])

In [98]:
import json

In [99]:
stores = r.json()["store"]

In [100]:
stores

[{'IsMobileShoppingEnabled': 'Y',
  'IsMobilePaymentEnabled': 'N',
  'unitid': 316,
  'streetAddress': '2191 Columbia Avenue West',
  'city': 'Battle Creek',
  'state': 'MI',
  'zip': '49015-2847',
  'latitude': 42.29455,
  'longitude': -85.23586,
  'storeHours': 'Open 24 hrs a day, 364 days a year.',
  'storeName': '2191 Columbia Avenue West, Battle Creek, MI 49015-2847',
  'milesFrom': 7.15879690243091,
  'storeShortName': 'W. Columbia Ave. - Battle Creek',
  'storePhoneNumber': '2699683371',
  'pharmacyPhone': '2699660310',
  'isPharmacy': False,
  'pharmacyHours': [{'dayHours': '8:00 AM - 9:00 PM', 'dayOfTheWeek': 'M-F'},
   {'dayHours': '9:00 AM - 7:00 PM', 'dayOfTheWeek': 'Sat'},
   {'dayHours': '10:00 AM - 6:00 PM', 'dayOfTheWeek': 'Sun'}]},
 {'IsMobileShoppingEnabled': 'Y',
  'IsMobilePaymentEnabled': 'N',
  'unitid': 21,
  'streetAddress': '5800 Gull Rd',
  'city': 'Kalamazoo',
  'state': 'MI',
  'zip': '49048',
  'latitude': 42.32322,
  'longitude': -85.51544,
  'storeHours':

In [94]:
with open("meijer_stores.json", "w") as fid:
    json.dump(stores, fid)

In [95]:
with open("meijer_stores.json", "r") as fid:
    stores2 = json.load(fid)

In [96]:
stores2

[{'IsMobileShoppingEnabled': 'Y',
  'IsMobilePaymentEnabled': 'N',
  'unitid': 316,
  'streetAddress': '2191 Columbia Avenue West',
  'city': 'Battle Creek',
  'state': 'MI',
  'zip': '49015-2847',
  'latitude': 42.29455,
  'longitude': -85.23586,
  'storeHours': 'Open 24 hrs a day, 364 days a year.',
  'storeName': '2191 Columbia Avenue West, Battle Creek, MI 49015-2847',
  'milesFrom': 7.15879690243091,
  'storeShortName': 'W. Columbia Ave. - Battle Creek',
  'storePhoneNumber': '2699683371',
  'pharmacyPhone': '2699660310',
  'isPharmacy': False,
  'pharmacyHours': [{'dayHours': '8:00 AM - 9:00 PM', 'dayOfTheWeek': 'M-F'},
   {'dayHours': '9:00 AM - 7:00 PM', 'dayOfTheWeek': 'Sat'},
   {'dayHours': '10:00 AM - 6:00 PM', 'dayOfTheWeek': 'Sun'}]},
 {'IsMobileShoppingEnabled': 'Y',
  'IsMobilePaymentEnabled': 'N',
  'unitid': 21,
  'streetAddress': '5800 Gull Rd',
  'city': 'Kalamazoo',
  'state': 'MI',
  'zip': '49048',
  'latitude': 42.32322,
  'longitude': -85.51544,
  'storeHours':